src/NovelRT/Experimental/Graphics/Vulkan/VulkanGraphicsBuffer.cpp

```diff
@@ -0,0 +1,237 @@
+        VkDeviceMemory vulkanDeviceMemory = GetBlock()->GetVulkanDeviceMemory();
+
+        void* pDestination = nullptr;
+
+        VkResult mapMemoryResult =
+            vkMapMemory(vulkanDevice, vulkanDeviceMemory, GetOffset(), GetSize(), 0, &pDestination);
+
+        if (mapMemoryResult != VK_SUCCESS)
+        {
+            // TODO: Make a real exception.
+            throw std::runtime_error("Failed to map Vulkan memory to the CPU!");
+        }
+
+        uint64_t nonCoherentAtomSize =
+            device->GetAdapter()->GetVulkanPhysicalDeviceProperties().limits.nonCoherentAtomSize;
+
+        size_t offset = GetOffset() + readRangeOffset;
+        size_t size = (readRangeLength + nonCoherentAtomSize - 1) & ~(nonCoherentAtomSize - 1);
+
+        VkMappedMemoryRange mappedMemoryRange{};
+        mappedMemoryRange.sType = VK_STRUCTURE_TYPE_MAPPED_MEMORY_RANGE;
+        mappedMemoryRange.memory = vulkanDeviceMemory;
+        mappedMemoryRange.offset = offset;
+        mappedMemoryRange.size = size;
+
+        VkResult invalidateMappedMemoryRangesResult =
+            vkInvalidateMappedMemoryRanges(device->GetVulkanDevice(), 1, &mappedMemoryRange);
+
+        if (invalidateMappedMemoryRangesResult != VK_SUCCESS)
+        {
+            // TODO: Make a real exception.
+            throw std::runtime_error("Failed to map the Vulkan memory for read only!");
+        }
+
+        return (reinterpret_cast<uint8_t*>(pDestination) + readRangeOffset);
+    }
+
+    void VulkanGraphicsBuffer::Unmap()
+    {
+        std::shared_ptr<VulkanGraphicsDevice> device = GetDevice();
+
+        VkDevice vulkanDevice = device->GetVulkanDevice();
+        VkDeviceMemory vulkanDeviceMemory = GetBlock()->GetVulkanDeviceMemory();
+
+        vkUnmapMemory(vulkanDevice, vulkanDeviceMemory);
+    }
+
+    void VulkanGraphicsBuffer::UnmapAndWrite()
+    {
+        std::shared_ptr<VulkanGraphicsDevice> device = GetDevice();
+
+        VkDevice vulkanDevice = device->GetVulkanDevice();
+        VkDeviceMemory vulkanDeviceMemory = GetBlock()->GetVulkanDeviceMemory();
+
+        uint64_t nonCoherentAtomSize =
+            device->GetAdapter()->GetVulkanPhysicalDeviceProperties().limits.nonCoherentAtomSize;
+
+        size_t offset = GetOffset();
+        size_t size = (GetSize() + nonCoherentAtomSize - 1) & ~(nonCoherentAtomSize - 1);
+
+        VkMappedMemoryRange mappedMemoryRange{};
+        mappedMemoryRange.sType = VK_STRUCTURE_TYPE_MAPPED_MEMORY_RANGE;
+        mappedMemoryRange.memory = vulkanDeviceMemory;
+        mappedMemoryRange.offset = offset;
+        mappedMemoryRange.size = size;
+
+        VkResult flushMappedMemoryRangesResult = vkFlushMappedMemoryRanges(vulkanDevice, 1, &mappedMemoryRange);
+
+        if (flushMappedMemoryRangesResult != VK_SUCCESS)
+        {
+            // TODO: Make a real exception.
+            throw std::runtime_error("Failed to flush the written changes to the Vulkan memory!");
```

Comment:
I'm a little jelly now - I don't think I can get a real exception either... can I?

Reply:
This is probably another design discussion ticket to see what would be appropriate.